### DISCUSSION TITLE
Newlines in some constructs within interpolated regular string literals

### DISCUSSION THREAD

**Discussion by svick** [dotnet/csharplang#1513] [Q&A]
When looking at [the spec for interpolated string literals](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#interpolated-string-literals) because of https://github.com/dotnet/roslyn/issues/26814, I found some discrepancies: as far as I can tell, the lexical grammar says that in some specific cases, newlines should be allowed within interpolated regular (i.e. non-verbatim) string literals.

Specifically, those cases are when the `single_regular_balanced_text_character` rule is *not* used and instead one of the rules `delimited_comment`, `string_literal` or `interpolated_string_literal` is used.

I think the intention is that newlines should not be allowed in those cases and the compiler does not allow them.

Consider this expression:

```c#
$"{@"
"}"
```

I believe the lexical grammar in the spec says this should be parsed as:

* `interpolated_string_literal`
* `'$' interpolated_regular_string_literal`
* `'$' interpolated_regular_string_start interpolated_regular_string_literal_body interpolated_regular_string_end`
* `'$' '"' interpolated_regular_string_character* '{' interpolated_regular_string_literal_body interpolated_regular_string_end`
* `'$' '"' '{' interpolated_regular_string_literal_body interpolated_regular_string_end`
* `'$' '"' '{' regular_balanced_text interpolated_regular_string_end`
* `'$' '"' '{' regular_balanced_text_part+ interpolated_regular_string_end`
* `'$' '"' '{' regular_balanced_text_part interpolated_regular_string_end`
* `'$' '"' '{' string_literal interpolated_regular_string_end`
* `'$' '"' '{' verbatim_string_literal interpolated_regular_string_end`
* `'$' '"' '{' '@"' verbatim_string_literal_character* '"' interpolated_regular_string_end`
* `'$' '"' '{' '@"' verbatim_string_literal_character '"' interpolated_regular_string_end`
* `'$' '"' '{' '@"' '<Line feed character (U+000A)>' '"' interpolated_regular_string_end`
* `'$' '"' '{' '@"' '<Line feed character (U+000A)>' '"' interpolation_format? '}' interpolated_regular_string_characters_after_brace? '"'`
* `'$' '"' '{' '@"' '<Line feed character (U+000A)>' '"' '}' interpolated_regular_string_characters_after_brace? '"'`
* `'$' '"' '{' '@"' '<Line feed character (U+000A)>' '"' '}' '"'`

Thus, according to my reading of the spec, this expression should be valid, even though it goes against the presumed intention of interpolated regular string literals and even though [the compiler does not allow it](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEBMACAwhg3gWACgMSMUAWDAWQAoBKPI05jAfQwF4MASAIlxQpeRXgF9eAbialRRUUA===).

Is this really a bug in the spec? If not, where am I wrong?

**Comment by Joe4evr** [comment]
Should maybe be noted: [adding one more `@` character solves it](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEBMACAwhg3gWACgMSMUAWDAWQAoBKPI05jAfQwF4MASFFAIlx9+RfgF9+AbiakxRMUA==).

**Comment by svick** [comment]
@Joe4evr Right, that turns the expression into an interpolated verbatim string literal, where newlines are allowed.

**Comment by jcouv** [comment]
Tagging @gafter

**Comment by Unknown6656** [comment]
Also consider allowing:
```csharp
IEnumerable <...> list = ...;
string s = $"This is my enumeration: {string.Join(", ", from x in list
                                                        where x.SomeCondition
                                                        select Process(x))}";
```

**Comment by CyrusNajmabadi** [comment]
@svick My reading of the spec agrees with yours.
